Samples/ios/SampleApp/SampleAppiOS/SampleMSALUtil.m

```diff
@@ -0,0 +1,217 @@
+    // image from another device. In this case it is best to detect that case and clean up local state.
+    if (!user && [localError.domain isEqualToString:MSALErrorDomain] && localError.code == MSALErrorUserNotFound)
+    {
+        [self cleanupLocalState];
+    }
+    
+    // NSError ** is traditionally an optional parameters, so we should check to make sure it is not nil before
+    // filling it.
+    if (error)
+    {
+        *error = localError;
+    }
+    
+    return user;
+}
+
++ (void)signInUser:(void (^)(MSALUser *user, NSString *token, NSError *error))signInBlock
+{
+    MSALPublicClientApplication *application = [self clientApplication];
+    
+    // When signing in a user for the first time we acquire a token without providing
+    // a user object. If you've previously asked the user for an email address,
+    // or phone number you can provide that as a "login hint."
+    [application acquireTokenForScopes:@[@"User.Read"]
+                       completionBlock:^(MSALResult *result, NSError *error)
+    {
+        if (error)
+        {
+            signInBlock(nil, nil, error);
+            return;
+        }
+        
+        // In the initial acquire token call we'll want to look at the user object
+        // that comes back in the result.
+        MSALUser *user = result.user;
+        
+        // The userIdentifier in the MSALUser is the key to retrieve this user from
+        // the cache in the future. Save this piece of information in a place you can
+        // easily retrieve in your app. In this case we're going to store it in
+        // NSUserDefaults.
+        [[NSUserDefaults standardUserDefaults] setValue:user.userIdentifier forKey:CURRENT_USER_KEY];
+        
+        signInBlock(user, result.accessToken, error);
+    }];
+}
+
++ (void)acquireTokenSilentForCurrentUser:(void (^)(NSString *token, NSError *error))acquireTokenBlock
+{
+    MSALPublicClientApplication *application = [self clientApplication];
+    
+    NSError *error = nil;
+    MSALUser *currentUser = [self currentUser:&error];
+    if (!currentUser)
+    {
+        acquireTokenBlock(nil, error);
+        return;
+    }
+    
+    // Depending on how this user has been used with this application before it is possible for there to be multiple
+    // tokens of varying authorities for this user in the cache. Because we are trying to get a token specifically
+    // for graph in this sample it's best to specify the user's home authority to remove any possibility of there
+    // being any ambiquity in the cache lookup.
+    NSString *homeAuthority = [NSString stringWithFormat:@"https://login.microsoftonline.com/%@", currentUser.utid];
+    
+    
+    [application acquireTokenSilentForScopes:@[@"User.Read"]
+                                        user:currentUser
+                                   authority:homeAuthority
+                                forceRefresh:NO
+                               correlationId:nil
```

Comment:
naively, looking through just this code with fresh eyes, I'm not sure what correlationId is supposed to represent. it may be worth a comment why you're passing in nil here, and in what cases someone'd want to pass in something different.